### ISSUE TITLE
There isn't a way to obtain the root world directory until worlds load.

### ISSUE THREAD

**Issue by gravityfox** [SpongePowered/SpongeCommon#292]
Calling Game#getSavesDirectory() just returns the root directory of the server.

It would be nice to have a method that gives me the root world directory before the world loads, so that I can actually do some meaningful loading of data that has to happen before the worlds load.

The way I have been doing it is to wait for all the worlds to load first, and then get the name of the default world and using that for folder access. It's messy and I really don't like it. I also don't want to make a separate folder next to the default world folder because the data i'm trying to load is specific to the world, it just has to be loaded beforehand.

**Comment by gravityfox** [comment]
Also it appears impossible to actually obtain the root world directory if someone makes a new world on the client in single player, Since the world save name is different than the dimension name.

**Comment by Zidane** [comment]
I'll look into this @gravityfox .

**Comment by bloodmc** [comment]
@gravityfox That is the whole point of ConstructWorldEvent which happens before a world loads.
Calling getSavesDirectory will be available after overworld loads which is always the first world to load.
You can also use https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/Server.java#L130
to construct the root directory easily.

**Comment by gravityfox** [comment]
I understood that bit, and that is how I am currently getting the world directory. However I have data that needs to be loaded before the first world is even constructed, and I want to be able to store that data into the root world folder. The root folder should be just loaded from properties.

**Comment by bloodmc** [comment]
I don't understand how this is an issue with the ConstructWorldEvent. Please post the code you are trying to get working.

**Comment by gravityfox** [comment]
My code isn't consolidated into one place unfortunately.

I can try moving some code toward the end of world loading and use world construction to load the initial data, but that just seems disorderly.

Also, does getSavesDirectory() return different things pre and post loading of the overworld?

**Comment by JBYoshi** [comment]
@gravityfox At what point are you trying to get the per-world data?

**Comment by gravityfox** [comment]
So I have data pertaining to the world save in general, and then i have data that is specific to each dimension. I need the general data to load BEFORE the world specific data, and I would like to load the general data before any of the world dimensions actually start loading.

**Comment by bloodmc** [comment]
You need to provide EXACTLY what you are trying to do. Stating "general data" doesn't help one bit.

**Comment by gravityfox** [comment]
I am trying to load a database file that resides in the root world directory. Then i need to load a database file for every dimension. The database files for the dimensions need to be loaded after the database in the root world directory, or else they data that they reference won't be available, and loading out of order causes issues. I also need to support single player mode, where getting the saves directory and appending the world name won't cut it.

**Comment by ST-DDT** [comment]
And why exactly must it be loaded before the worlds are constructed?

PS: The javadocs on ConstructWorldEvent and LoadWorldEvent should be improved to state when they are called. 

Example, may be wrong:  
Construct before anything is created only the WorldSettings have been read..  
Load is called (shortly) after ConstructEv3nt after the world has been instantiated and all settings are loaded but before the first chunks are loaded.

**Comment by gravityfox** [comment]
The loading process for each dimension refers to a global set of data that should still be unique to the worlds in general. It doesn't feel appropriate the load the global data during a ConstructWorldEvent because that data isn't tied to any particular world.

**Comment by bloodmc** [comment]
OK, I'll fix getSavesDirectory tomorrow.

**Comment by ST-DDT** [comment]
You could also listen to GameStateEvents.
[GameState](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/GameState.java)
Especially the game INITIALIZATION step and SERVER_ABOUT_TO_START step may be the events you could use instead.

**Comment by gravityfox** [comment]
That's what I want to do. However with the current API and implementation, you can't get the root world directory until the overworld has been constructed.

INITIALIZATION and SERVER_ABOUT_TO_START happen before that.

This is literally why I made this issue in the first place.

**Comment by ST-DDT** [comment]
if the global data are not bound to a world they should not be stored in one.

**Comment by gravityfox** [comment]
But that data IS important to the WORLDS as a WHOLE.
It just isn't bound to any singular world.

**Comment by ST-DDT** [comment]
if you want to package your worlds as a single folder and make it available for download to play your Game/RPG then at least IMO they are world dependend. And than you should use the ConstructWorldEvent. You can even use that check to validate that the worlds are loaded in the correct order. As an alternative you could load the global database with the first world that needs it. Just check the parent dir of the given world for the global database.

**Comment by gravityfox** [comment]
And then you have to check for default worlds and check if it's single player or multiplayer.

Now I get this messy code that I started with where things load out of order and check global variables for load status.

The problem isn't that it doesn't work. It's that it doesn't work in the most logically reasonable manner.

**Comment by ST-DDT** [comment]
Why the check for SP or MP?
You will always need complex code to avoid odering/initialization issues.
Imagine somebody uses your special worlds with a MultiWorld plugin. Then the precious global world data won't be in the default world anymore...

**Comment by gravityfox** [comment]
The consensus of Sponge was to keep the default minecraft world directory structure.

If a plugin decides to mess with that, I will expect it to also redirect a method like 
getDefaultWorldDirectory() to a new, meaningful location.

**Comment by ST-DDT** [comment]
The MultiWorld plugin does not alter the server behaviour at all.
It just loads a different default world and your world as an extesion world.

You either habe to introduce structural limitations or deal with every possible case in case..

Are you using the default hierachical world setup or a flat world setup?
In one case placing the global world data in the default world is enough and you can easily achieve loading the data if any world needs it/is loaded (either checking that or it parent folder for the data container). Or if you already habe multiple folders on the root level then you could just add one more.(with a const name)
IMO both solutions wont cause complex code at all. You could also combine both.

Hierarchical
* world
  * regions
  * myData.db
  * dimx
    * regions
  * dimy
    * regions

Flat
* world
  * regions
* myData.db
* worldX
  * regions
* worldY
  * regions

**Comment by gravityfox** [comment]
That is lazy resource loading. Something I SPECIFICALLY want to avoid.

**Comment by simon816** [comment]
Remember that the API doesn't define any particular method of world storage so generally speaking making assumptions on the structure of the implementation is a bad idea. I think at some point it could have been possible for world data to be stored in a mysql database and have no flatfile storage at all. Don't think the API allows this freedom anymore since `getSavesDirectory` was added

**Comment by Zidane** [comment]
@gravityfox 

Using ```getSavesDirectory()``` along with the newly added ```getDefaultWorldName()``` of ```Server``` will let you do what you want to do in ```GameAboutToStartServerEvent``` before worlds are constructed.

Don't do this before this point else it won't be the correct name.
